### PR TITLE
Complete type annotations of several files

### PR DESCRIPTION
**Pull request by jdufresne**

### CODE REVIEW

File: src/urllib3/_collections.py [hunk 1]
```diff
@@ -255,7 +255,7 @@ def __init__(self, headers: Optional[ValidHTTPHeaderSource] = None, **kwargs: st
         if kwargs:
             self.extend(kwargs)
 
-    def __setitem__(self, key: str, val: str) -> None:
+    def __setitem__(self, key: Union[bytes, str], val: str) -> None:
```
Comment (sethmlarson) [inline src/urllib3/_collections.py:258]: So this is true but I'm not sure if we should recommend it? Wasn't there some discussion about `bytes` headers somewhere, can't find it now.

File: test/contrib/test_socks.py [hunk 2]
```diff
@@ -294,16 +311,17 @@ def request_handler(listener):
...
+        def request_handler(listener: socket.socket) -> None:
             sock = listener.accept()[0]
 
             handler = handle_socks5_negotiation(sock, negotiate=False)
             addr, port = next(handler)
 
             assert addr == b"example.com"
             assert port == 80
-            handler.send(True)
+            with pytest.raises(StopIteration):
```
Comment (sethmlarson) [inline test/contrib/test_socks.py:323]: Why is this added?

**Review by sethmlarson — CHANGES_REQUESTED** [review]
This is really great, thanks for this! A few comments for you below:

File: src/urllib3/__init__.py [hunk 3]
```diff
@@ -22,9 +22,11 @@
 
 __author__ = "Andrey Petrov (andrey.petrov@shazow.net)"
 __license__ = "MIT"
-__version__ = __version__
 
 __all__ = (
+    "__author__",
```
[the remark this answers is not in the document]
Comment (jdufresne) [reply src/urllib3/__init__.py:27]: That makes sense! I added it here for the sake of `docs/conf.py`, but I think it would be better to continue excluding it from type checking.

File: src/urllib3/_collections.py (hunk 1 above)
Comment (jdufresne) [reply src/urllib3/_collections.py:258]: Is the preference to?

1. Add ignore comments to all uses
2. Only handle str, update tests, implementation and the type signature?

File: test/contrib/test_socks.py (hunk 2 above)
Comment (jdufresne) [reply test/contrib/test_socks.py:323]: The original implementation of `handle_socks*_negotiation` would yield multiple types of varying types. Sometimes a (addr, port) tuple and other times a bool. The bool looks like it was added to avoid handling a `StopIteration` exception.The yielded value is not actually used. We could define the generator to yield `Union[Tuple, bool]` however, this complicates the call sites as now the type checker will complain that `bool` is not iterable. I found it simpler to only yield a single type `(addr, port)` and in other code branches to stop the generator, hence `StopIteration`.

Is there a preference for a different approach?

File: src/urllib3/_collections.py (hunk 1 above)
Comment (jdufresne) [reply src/urllib3/_collections.py:258]: I went with option 2 in #2397 but am very open to discussing other options.

Comment (jdufresne) [reply src/urllib3/_collections.py]: I have reverted the changes in `src/urllib3/_collections.py`. They now continue to be typed for `str` only even though the implementation allows bytes. I have added `# type: ignore` comments to the necessary tests with a comment. Thanks for clarifying.

Comment (sethmlarson) [reply src/urllib3/_collections.py]: I appreciate you taking initiative here, we may end up deprecating using `bytes` for this case to push a little harder towards using our APIs as intended. We can handle that separately though.

File: test/contrib/test_socks.py (hunk 2 above)
Comment (sethmlarson) [reply test/contrib/test_socks.py:323]: Your approach is much better, thanks for making this change :+1:

**Review by sethmlarson — APPROVED** [review]
LGTM, thanks for all this great work :muscle: Is `jon.dufresne@gmail.com` the best way to reach you?

**Comment by jdufresne** [comment]
> Is jon.dufresne@gmail.com the best way to reach you?

Yup. :slightly_smiling_face:
